### PR TITLE
feat(sdl): Add SDL to public schema

### LINKED ISSUE #6248: DX(PrivateSchema): separate authorization from resolvers
Predecessor: #6228

Using the private schema as a test, how should we separate authorization and resolution?

By breaking them apart, we have a clear separation of concerns, which is good for scalability.
It also means we should be able to identify glaring holes in our permissions.

### Suggested Options

- GraphQL Shield: https://graphql-shield.vercel.app/
- Resolver composition: https://www.graphql-tools.com/docs/resolvers-composition
- Authz: https://www.the-guild.dev/blog/graphql-authz (probably my least favorite)

### Considerations

Some mutations may change the underlying authorization. If authorization is checked using a dataloader, then that dataloader would have to be cleared before returning the mutation resolution, otherwise it'll result in stale authorization.
For example, let's say I call the mutation `removeFromTeam` which requires the viewer to be on TeamX. The authorizer uses a dataloader to see if an active TeamMember exists. Then, the resolver inactivates that team member & returns the team member. if that team member gets resolved using the dataloader, it'll still show up as active. 

### Background
This was the challenge problem for backend devs: https://www.notion.so/parabol/Back-end-Technical-Problem-a9d281e0396b41be85865e2b3a8461b4

We also talk about it here: https://github.com/ParabolInc/parabol/issues/4540#issuecomment-853465969

AC
- [ ] Create a 1 or many draft PRs that show what the code would look like using each option.
- [ ] If there are other options to consider, bring them up & weigh the pros/cons

Estimate: 8 hours

### PR DESCRIPTION
**Pull request by mattkrick**
fix #6248 

The SDL pattern on the private schema works pretty well.
Next step is to support building things in the new way on the public schema.

To support incremental transitions, i created a `graphql/public` directory that will hold all the new patterns.
We can then move e.g. `graphql/queries` over to `graphql/public/queries` one by one whenever we prioritize it.

How to test: 
- About 40 files are just removing the old authorization from the resolvers. You can safely ignore those changes

### CODE REVIEW

File: packages/server/graphql/mutations/helpers/autoGroup/addLemmaToEntities.ts [hunk 1]
```diff
@@ -50,13 +49,11 @@ interface EntityWithoutLemma {
...
-): GoogleAnalyzedEntity[] => {
-  if (!syntax) return entities
-  const {tokens} = syntax
-  if (!Array.isArray(tokens)) return entities
+) => {
+  const tokens = syntax?.tokens
   return entities.map((entity) => ({
     ...entity,
-    lemma: addLemmaToEntity(entity.name, tokens)
+    lemma: tokens ? addLemmaToEntity(entity.name, tokens) : ''
```
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/autoGroup/addLemmaToEntities.ts:56]: the new typesafety discovered this had to be non-null

File: packages/server/graphql/public/queries/getDemoEntities.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import getReflectionEntities from '../../mutations/helpers/getReflectionEntities'
+import {QueryResolvers} from '../resolverTypes'
+
+const getDemoEntities: QueryResolvers['getDemoEntities'] = async (_source, {text}) => {
```
Comment (mattkrick) [inline packages/server/graphql/public/queries/getDemoEntities.ts:4]: i had to codemod 1 query to populate the query directory

**Review by jmtaber129 — APPROVED** [review]

**Comment by BartoszJarocki** [comment]
@mattkrick are you planning to update the `newMutation` script to generate code in a new manner? should we create a task for it? great work here!

File: packages/server/graphql/composeResolvers.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  [TypeName: string]: any
+}
+
+interface PermissionMap {
+  [TypeName: string]: {
+    [FieldName: string]: ShieldRule
+  }
+}
+
+const composeResolvers = <T extends ResolverMap>(resolverMap: T, permissionMap: PermissionMap) => {
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/graphql/composeResolvers.ts:61]: the function has been made pure ⛪

File: packages/server/graphql/public/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+If you need to call a service, use the service Manager (e.g. [AtlassianManager](../../utils/AtlassianServerManager.ts))
+Publish the event to a subscription channel to alert others users who are interested in the event
+Return the smallest possible object you can. For example, instead of returning an entire meeting object, return `{meetingId}`
+
+3. Write permissions
+Add rules to decide who can call the mutation in the [permissions](./permissions.ts)
+You may also decide to validate arguments in the permissions, too.
+
+4. Write the mutation payload source
+Create a type with the same shape as the return value of your mutation.
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/graphql/public/README.md:25]: ah! this is good signal.
the "mutation payload" is the shape of the output. the "mutation payload source" is the shape of the input.

e.g.
 
```ts
source = {meetingId}
mutationPayload = {meeting: db.load(meetingId)}
```

i'll rewrite

File: packages/server/utils/updateGQLSchema.ts [hunk 5]
```diff
@@ -3,20 +3,19 @@ import {printSchema} from 'graphql'
...
-  'packages/server/graphql/private/schema.graphql'
-)
 const updateGQLSchema = async () => {
-  const nextSchema = printSchema(publicSchema)
-  const nextPrivateSchema = printSchema(privateSchema)
-  await Promise.all([write(schemaPath, nextSchema), write(privateSchemaPath, nextPrivateSchema)])
+  const write = promisify(fs.writeFile)
+  const GQL_ROOT = path.join(__PROJECT_ROOT__, 'packages/server/graphql')
+  const publicSchemaPath = path.join(GQL_ROOT, 'public/schema.graphql')
+  const privateSchemaPath = path.join(GQL_ROOT, 'private/schema.graphql')
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/utils/updateGQLSchema.ts:14]: good question, comments added

File: scripts/codeshift/README.md [hunk 6]
```diff
@@ -1,21 +0,0 @@
...
-- Adding the mutation to the client-side subscription
-- A server-side mutation
-- A server-side mutation payload
-- Adding the server-side mutation to the rootMutations
-- Adding the server-side mutation payload to the subscription
-
-To speed things up, we wrote a script to do this automatically.
-
-```sh
-yarn newMutation <<mutationName>> -s <<subscriptionName>>
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply scripts/codeshift/README.md:15]: definitely should be modified, i'll write a new issue to tackle that work

**Comment by mattkrick** [comment]
@BartoszJarocki yep, i gotta still do that. tracking it in #6324

**Pull request merged**